File: backend/src/modules/auth/infrastructure/supabase/supabase-auth-client.ts
import { Injectable } from '@nestjs/common';
import { getSupabaseClient } from '../../../../shared/supabase/client';
import { ConfigService } from '@nestjs/config';
import { SupabaseAuthClient, AuthTokens, User, UserId, UserRole } from '../../domain';

@Injectable()
export class SupabaseAuthClientImpl implements SupabaseAuthClient {
  constructor(private readonly configService: ConfigService) {}

  private get client() {
    return getSupabaseClient(this.configService);
  }

  async signInWithEmail(email: string, password: string): Promise<AuthTokens> {
    const { data, error } = await this.client.auth.signInWithPassword({
      email,
      password,
    });

    if (error || !data.session) {
      throw new Error(`Sign in failed: ${error?.message}`);
    }

    return {
      accessToken: data.session.access_token,
      refreshToken: data.session.refresh_token,
      expiresIn: data.session.expires_in,
    };
  }

  async signUpWithEmail(email: string, password: string, metadata: { nombre: string }): Promise<AuthTokens> {
    const { data, error } = await this.client.auth.signUp({
      email,
      password,
      options: {
        data: metadata,
      },
    });

    if (error || !data.session) {
      throw new Error(`Sign up failed: ${error?.message}`);
    }

    return {
      accessToken: data.session.access_token,
      refreshToken: data.session.refresh_token,
      expiresIn: data.session.expires_in,
    };
  }

  async refreshAccessToken(refreshToken: string): Promise<AuthTokens> {
    const { data, error } = await this.client.auth.refreshSession({
      refresh_token: refreshToken,
    });

    if (error || !data.session) {
      throw new Error(`Token refresh failed: ${error?.message}`);
    }

    return {
      accessToken: data.session.access_token,
      refreshToken: data.session.refresh_token,
      expiresIn: data.session.expires_in,
    };
  }

  async signOut(): Promise<void> {
    await this.client.auth.signOut();
  }

  async getUser(accessToken: string): Promise<User | null> {
    const { data, error } = await this.client.auth.getUser(accessToken);

    if (error || !data.user) {
      return null;
    }

    return this.mapToUser(data.user);
  }

  async updateUserRole(userId: string, role: UserRole): Promise<void> {
    const { error } = await this.client.auth.admin.updateUserById(userId, {
      user_metadata: { role },
    });

    if (error) {
      throw new Error(`Failed to update user role: ${error.message}`);
    }
  }

  private mapToUser(supabaseUser: any): User {
    return {
      id: supabaseUser.id as UserId,
      email: supabaseUser.email!,
      role: (supabaseUser.user_metadata?.role as UserRole) || 'estudiante',
      nombre: supabaseUser.user_metadata?.nombre || supabaseUser.email!,
      activo: true,
      createdAt: new Date(supabaseUser.created_at),
      updatedAt: new Date(supabaseUser.updated_at || supabaseUser.created_at),
    };
  }
}

File: frontend/lib/features/auth/infrastructure/supabase_auth_client.dart
// lib/features/auth/infrastructure/supabase_auth_client.dart

import 'package:supabase_flutter/supabase_flutter.dart' as supabase;
import 'package:either_dart/either.dart';
import '../domain/entities.dart';
import '../domain/repositories.dart';
import '../../../core/storage/secure_storage.dart';

class SupabaseAuthClient implements AuthRepository {
  final supabase.SupabaseClient _client = supabase.Supabase.instance.client;
  final SecureStorage _secureStorage;

  SupabaseAuthClient(this._secureStorage);

  @override
  Future<Either<String, AuthResult>> signIn(String email, String password) async {
    try {
      final response = await _client.auth.signInWithPassword(
        email: email,
        password: password,
      );

      if (response.session == null || response.user == null) {
        return const Left('Credenciales inválidas');
      }

      final tokens = AuthTokens(
        accessToken: response.session!.accessToken,
        refreshToken: response.session!.refreshToken ?? '',
        expiresIn: response.session!.expiresIn ?? 0,
      );

      await _secureStorage.setAccessToken(tokens.accessToken);
      await _secureStorage.setRefreshToken(tokens.refreshToken);
      await _secureStorage.setUserId(response.user!.id);

      final user = User.fromJson({
        'id': response.user!.id,
        'email': response.user!.email!,
        'role': response.user!.userMetadata?['role'] ?? 'estudiante',
        'nombre': response.user!.userMetadata?['nombre'] ?? response.user!.email!,
        'activo': true,
      });

      return Right(AuthResult(user: user, tokens: tokens));
    } on supabase.AuthException catch (e) {
      return Left(e.message);
    } catch (e) {
      return Left('Error al iniciar sesión: $e');
    }
  }

  @override
  Future<Either<String, AuthResult>> signUp(String email, String password, String nombre) async {
    try {
      final response = await _client.auth.signUp(
        email: email,
        password: password,
        data: {'nombre': nombre},
      );

      if (response.session == null || response.user == null) {
        return const Left('Error al registrar usuario');
      }

      final tokens = AuthTokens(
        accessToken: response.session!.accessToken,
        refreshToken: response.session!.refreshToken ?? '',
        expiresIn: response.session!.expiresIn ?? 0,
      );

      await _secureStorage.setAccessToken(tokens.accessToken);
      await _secureStorage.setRefreshToken(tokens.refreshToken);
      await _secureStorage.setUserId(response.user!.id);

      final user = User.fromJson({
        'id': response.user!.id,
        'email': response.user!.email!,
        'role': response.user!.userMetadata?['role'] ?? 'estudiante',
        'nombre': response.user!.userMetadata?['nombre'] ?? nombre,
        'activo': true,
      });

      return Right(AuthResult(user: user, tokens: tokens));
    } on supabase.AuthException catch (e) {
      return Left(e.message);
    } catch (e) {
      return Left('Error al registrarse: $e');
    }
  }

  @override
  Future<Either<String, AuthTokens>> refreshToken(String refreshToken) async {
    try {
      final response = await _client.auth.refreshSession(refreshToken);
      
      if (response.session == null) {
        return const Left('No se pudo refrescar la sesión');
      }

      final tokens = AuthTokens(
        accessToken: response.session!.accessToken,
        refreshToken: response.session!.refreshToken ?? '',
        expiresIn: response.session!.expiresIn ?? 0,
      );

      await _secureStorage.setAccessToken(tokens.accessToken);
      await _secureStorage.setRefreshToken(tokens.refreshToken);

      return Right(tokens);
    } on supabase.AuthException catch (e) {
      return Left(e.message);
    } catch (e) {
      return Left('Error al refrescar token: $e');
    }
  }

  @override
  Future<void> signOut() async {
    await _client.auth.signOut();
    await _secureStorage.clearAuth();
  }

  @override
  Future<Either<String, User>> getCurrentUser() async {
    try {
      final user = _client.auth.currentUser;
      if (user == null) {
        return const Left('No hay usuario autenticado');
      }

      return Right(User.fromJson({
        'id': user.id,
        'email': user.email!,
        'role': user.userMetadata?['role'] ?? 'estudiante',
        'nombre': user.userMetadata?['nombre'] ?? user.email!,
        'activo': true,
      }));
    } catch (e) {
      return Left('Error al obtener usuario: $e');
    }
  }
}